File: android/src/main/kotlin/br/com/cpndntech/google_maps_plus/MapObjectsManager.kt
package br.com.cpndntech.google_maps_plus

import android.animation.ValueAnimator
import android.os.Handler
import android.os.Looper
import android.view.animation.LinearInterpolator
import com.google.android.gms.maps.CameraUpdateFactory
import com.google.android.gms.maps.GoogleMap
import com.google.android.gms.maps.model.*

class MapObjectsManager(private val googleMap: GoogleMap, private val density: Float) {
    private val mainHandler = Handler(Looper.getMainLooper())
    
    val markers = mutableMapOf<String, Marker>()
    val circles = mutableMapOf<String, Circle>()
    val polylines = mutableMapOf<String, Polyline>()
    val polygons = mutableMapOf<String, Polygon>()
    
    private val animators = mutableMapOf<String, ValueAnimator>()
    private val pendingIcons = mutableMapOf<String, BitmapDescriptor>()
    
    var defaultSpeed: Double = 60.0
    var maxAnimationDuration: Long = 10000L
    var followedMarkerId: String? = null
    var followEnabled = true

    // MARKERS
    fun addMarker(m: Map<String, Any>) {
        addMarkerInternal(m)
    }

    private fun addMarkerInternal(m: Map<String, Any>) {
        val id = m["markerId"] as? String ?: return
        val position = Convert.toLatLng(m["position"]) ?: return
        val rotation = (m["rotation"] as? Number)?.toFloat() ?: 0.0f
        
        val existing = markers[id]
        if (existing != null) {
            animateMarkerMoveInternal(existing, id, position, rotation)
            Convert.interpretMarker(m, existing)
            
            val descriptor = Convert.toBitmapDescriptor(m["icon"])
            if (descriptor != null) {
                if (animators.containsKey(id)) {
                    pendingIcons[id] = descriptor
                } else {
                    existing.setIcon(descriptor)
                }
            }
        } else {
            val options = Convert.toMarkerOptions(m)
            googleMap.addMarker(options)?.let { 
                it.tag = id
                markers[id] = it 
            }
        }
    }

    fun moveMarker(id: String, lat: Double, lng: Double, rotation: Float) {
        markers[id]?.let { animateMarkerMoveInternal(it, id, LatLng(lat, lng), rotation) }
    }

    fun updateMarkerIcon(id: String, iconData: Any) {
        mainHandler.post {
            val descriptor = Convert.toBitmapDescriptor(iconData) ?: return@post
            if (animators.containsKey(id)) {
                pendingIcons[id] = descriptor
            } else {
                markers[id]?.setIcon(descriptor)
            }
        }
    }

    fun removeMarker(id: String) {
        mainHandler.post {
            animators[id]?.cancel()
            animators.remove(id)
            pendingIcons.remove(id)
            markers[id]?.remove()
            markers.remove(id)
        }
    }

    fun applyMarkerUpdates(updates: MapObjectUpdates<MarkerData>) {
        mainHandler.post {
            updates.toRemove.forEach { removeMarker(it) }
            (updates.toAdd + updates.toChange).forEach { addMarkerInternal(it.data) }
        }
    }

    private fun animateMarkerMoveInternal(marker: Marker, id: String, targetPosition: LatLng, rotation: Float) {
        val currentPosition = marker.position
        val results = FloatArray(1)
        android.location.Location.distanceBetween(
            currentPosition.latitude, currentPosition.longitude,
            targetPosition.latitude, targetPosition.longitude,
            results
        )
        val distanceInMeters = results[0]

        if (distanceInMeters < 0.1) {
            marker.position = targetPosition
            marker.rotation = rotation
            if (id == followedMarkerId && followEnabled) {
                googleMap.moveCamera(CameraUpdateFactory.newLatLng(targetPosition))
            }
            return
        }

        val projection = googleMap.projection
        val visibleBounds = projection.visibleRegion.latLngBounds
        val expandedBounds = expandBounds(visibleBounds, 1.2)

        if (!expandedBounds.contains(currentPosition) && !expandedBounds.contains(targetPosition)) {
            marker.position = targetPosition
            marker.rotation = rotation
            return
        }

        val durationMs = ((distanceInMeters / defaultSpeed * 1000).toLong()).coerceIn(300L, maxAnimationDuration)
        animators[id]?.cancel()

        val animator = ValueAnimator.ofFloat(0f, 1f)
        animator.duration = durationMs
        animator.interpolator = LinearInterpolator()

        val startRotation = marker.rotation
        var deltaRotation = rotation - startRotation
        if (deltaRotation > 180f) deltaRotation -= 360f
        if (deltaRotation < -180f) deltaRotation += 360f

        animator.addUpdateListener { anim ->
            val t = anim.animatedFraction
            val animLat = currentPosition.latitude + (targetPosition.latitude - currentPosition.latitude) * t
            val animLng = currentPosition.longitude + (targetPosition.longitude - currentPosition.longitude) * t
            val pos = LatLng(animLat, animLng)
            marker.position = pos
            marker.rotation = startRotation + deltaRotation * t
            
            if (id == followedMarkerId && followEnabled) {
                googleMap.moveCamera(CameraUpdateFactory.newLatLng(pos))
            }
        }

        animator.addListener(object : android.animation.AnimatorListenerAdapter() {
            override fun onAnimationEnd(animation: android.animation.Animator) {
                animators.remove(id)
                pendingIcons.remove(id)?.let { descriptor ->
                    markers[id]?.setIcon(descriptor)
                }
            }
        })

        animator.start()
        animators[id] = animator
    }

    private fun expandBounds(bounds: LatLngBounds, factor: Double): LatLngBounds {
        val center = bounds.center
        val latSpan = bounds.northeast.latitude - bounds.southwest.latitude
        val lngSpan = bounds.northeast.longitude - bounds.southwest.longitude
        val newNe = LatLng(center.latitude + latSpan * factor / 2.0, center.longitude + lngSpan * factor / 2.0)
        val newSw = LatLng(center.latitude - latSpan * factor / 2.0, center.longitude - lngSpan * factor / 2.0)
        return LatLngBounds(newSw, newNe)
    }

    // SHAPES (CIRCLES, POLYLINES, POLYGONS)
    fun applyCircleUpdates(updates: MapObjectUpdates<CircleData>) {
        mainHandler.post {
            updates.toRemove.forEach { removeCircleInternal(it) }
            (updates.toAdd + updates.toChange).forEach { addCircleInternal(it.data) }
        }
    }

    fun addCircle(c: Map<String, Any>) {
        addCircleInternal(c)
    }

    private fun addCircleInternal(c: Map<String, Any>) {
        val id = c["circleId"] as? String ?: return
        val existing = circles[id]
        if (existing != null) {
            Convert.interpretCircle(c, existing, density)
        } else {
            val options = Convert.toCircleOptions(c, density)
            googleMap.addCircle(options)?.let { 
                it.tag = id
                circles[id] = it 
            }
        }
    }

    fun removeCircle(id: String) {
        removeCircleInternal(id)
    }

    private fun removeCircleInternal(id: String) {
        circles[id]?.remove()
        circles.remove(id)
    }

    fun applyPolylineUpdates(updates: MapObjectUpdates<PolylineData>) {
        updates.toRemove.forEach { removePolylineInternal(it) }
        (updates.toAdd + updates.toChange).forEach { addPolylineInternal(it.data) }
    }

    fun addPolyline(p: Map<String, Any>) {
        addPolylineInternal(p)
    }

    private fun addPolylineInternal(p: Map<String, Any>) {
        val id = p["polylineId"] as? String ?: return
        val existing = polylines[id]
        if (existing != null) {
            Convert.interpretPolyline(p, existing, density)
        } else {
            val options = Convert.toPolylineOptions(p, density)
            googleMap.addPolyline(options)?.let { 
                it.tag = id
                polylines[id] = it 
            }
        }
    }

    fun removePolyline(id: String) {
        mainHandler.post { removePolylineInternal(id) }
    }

    private fun removePolylineInternal(id: String) {
        polylines[id]?.remove()
        polylines.remove(id)
    }

    fun applyPolygonUpdates(updates: MapObjectUpdates<PolygonData>) {
        updates.toRemove.forEach { removePolygonInternal(it) }
        (updates.toAdd + updates.toChange).forEach { addPolygonInternal(it.data) }
    }

    fun addPolygon(p: Map<String, Any>) {
        addPolygonInternal(p)
    }

    private fun addPolygonInternal(p: Map<String, Any>) {
        val id = p["polygonId"] as? String ?: return
        val existing = polygons[id]
        if (existing != null) {
            Convert.interpretPolygon(p, existing, density)
        } else {
            val options = Convert.toPolygonOptions(p, density)
            googleMap.addPolygon(options)?.let { 
                it.tag = id
                polygons[id] = it 
            }
        }
    }

    fun removePolygon(id: String) {
        mainHandler.post { removePolygonInternal(id) }
    }

    private fun removePolygonInternal(id: String) {
        polygons[id]?.remove()
        polygons.remove(id)
    }

    fun clearAll() {
        mainHandler.post {
            markers.values.forEach { it.remove() }
            markers.clear()
            circles.values.forEach { it.remove() }
            circles.clear()
            polylines.values.forEach { it.remove() }
            polylines.clear()
            polygons.values.forEach { it.remove() }
            polygons.clear()
            animators.values.forEach { it.cancel() }
            animators.clear()
            pendingIcons.clear()
        }
    }

    fun showMarkerInfoWindow(id: String) {
        mainHandler.post {
            markers[id]?.showInfoWindow()
        }
    }

    fun hideMarkerInfoWindow(id: String) {
        mainHandler.post {
            markers[id]?.hideInfoWindow()
        }
    }

    fun isMarkerInfoWindowShown(id: String): Boolean {
        return markers[id]?.isInfoWindowShown ?: false
    }
}

File: ios/Classes/MapObjectsManager.swift
import GoogleMaps
import Flutter

class MapObjectsManager: NSObject {
    private let mapView: GMSMapView
    private let registrar: FlutterPluginRegistrar
    
    var markers: [String: GMSMarker] = [:]
    var circles: [String: GMSCircle] = [:]
    var polylines: [String: GMSPolyline] = [:]
    var polygons: [String: GMSPolygon] = [:]
    
    var defaultSpeed: Double = 60.0
    var maxAnimationDuration: Double = 2.0
    var followedMarkerId: String? = nil
    var followEnabled = true

    init(mapView: GMSMapView, registrar: FlutterPluginRegistrar) {
        self.mapView = mapView
        self.registrar = registrar
    }

    func setAllMarkers(_ markersData: [[String: Any]]) {
        markers.values.forEach { $0.map = nil }
        markers.removeAll()
        for m in markersData { addMarker(m) }
    }

    func addMarker(_ data: [String: Any]) {
        guard let id = data["markerId"] as? String else { return }
        
        if let existing = markers[id] {
            if let pos = Convert.toCoordinate(data["position"]) {
                let rotation = Convert.toDouble(data["rotation"]) ?? 0.0
                moveMarkerInternal(marker: existing, id: id, lat: pos.latitude, lng: pos.longitude, rotation: rotation)
            }
            Convert.interpretMarker(data, marker: existing)
        } else {
            let marker = GMSMarker()
            Convert.interpretMarker(data, marker: marker)
            marker.userData = id
            marker.map = mapView
            markers[id] = marker
        }
    }

    func moveMarker(id: String, lat: Double, lng: Double, rotation: Double) {
        guard let marker = markers[id] else { return }
        moveMarkerInternal(marker: marker, id: id, lat: lat, lng: lng, rotation: rotation)
    }

    func updateMarkerIcon(id: String, iconData: Any) {
        guard let marker = markers[id] else { return }
        marker.icon = Convert.toIcon(iconData)
    }

    func removeMarker(_ id: String) {
        markers[id]?.map = nil
        markers.removeValue(forKey: id)
    }

    func applyMarkerUpdates(_ updates: MapObjectUpdates<MarkerData>) {
        updates.toRemove.forEach { removeMarker($0) }
        (updates.toAdd + updates.toChange).forEach { addMarker($0.data) }
    }

    private func moveMarkerInternal(marker: GMSMarker, id: String, lat: Double, lng: Double, rotation: Double) {
        let targetPosition = CLLocationCoordinate2D(latitude: lat, longitude: lng)
        let loc1 = CLLocation(latitude: marker.position.latitude, longitude: marker.position.longitude)
        let loc2 = CLLocation(latitude: lat, longitude: lng)
        let distance = loc1.distance(from: loc2)
        
        let duration = max(0.3, min(maxAnimationDuration, distance / defaultSpeed))
        
        CATransaction.begin()
        CATransaction.setAnimationDuration(duration)
        marker.position = targetPosition
        marker.rotation = rotation
        
        if id == followedMarkerId && followEnabled {
            mapView.animate(with: GMSCameraUpdate.setTarget(targetPosition))
        }
        CATransaction.commit()
    }

    // SHAPES
    func addCircle(_ data: [String: Any]) {
        guard let id = data["circleId"] as? String else { return }
        if let existing = circles[id] {
            Convert.interpretCircle(data, circle: existing)
        } else {
            let circle = GMSCircle()
            Convert.interpretCircle(data, circle: circle)
            circle.userData = id
            circle.map = mapView
            circles[id] = circle
        }
    }

    func removeCircle(_ id: String) {
        circles[id]?.map = nil
        circles.removeValue(forKey: id)
    }

    func applyCircleUpdates(_ updates: MapObjectUpdates<CircleData>) {
        updates.toRemove.forEach { removeCircle($0) }
        (updates.toAdd + updates.toChange).forEach { addCircle($0.data) }
    }

    func addPolyline(_ data: [String: Any]) {
        guard let id = data["polylineId"] as? String else { return }
        if let existing = polylines[id] {
            Convert.interpretPolyline(data, polyline: existing)
        } else {
            let polyline = GMSPolyline()
            Convert.interpretPolyline(data, polyline: polyline)
            polyline.userData = id
            polyline.map = mapView
            polylines[id] = polyline
        }
    }

    func removePolyline(_ id: String) {
        polylines[id]?.map = nil
        polylines.removeValue(forKey: id)
    }

    func applyPolylineUpdates(_ updates: MapObjectUpdates<PolylineData>) {
        updates.toRemove.forEach { removePolyline($0) }
        (updates.toAdd + updates.toChange).forEach { addPolyline($0.data) }
    }

    func addPolygon(_ data: [String: Any]) {
        guard let id = data["polygonId"] as? String else { return }
        if let existing = polygons[id] {
            Convert.interpretPolygon(data, polygon: existing)
        } else {
            let polygon = GMSPolygon()
            Convert.interpretPolygon(data, polygon: polygon)
            polygon.userData = id
            polygon.map = mapView
            polygons[id] = polygon
        }
    }

    func removePolygon(_ id: String) {
        polygons[id]?.map = nil
        polygons.removeValue(forKey: id)
    }

    func applyPolygonUpdates(_ updates: MapObjectUpdates<PolygonData>) {
        updates.toRemove.forEach { removePolygon($0) }
        (updates.toAdd + updates.toChange).forEach { addPolygon($0.data) }
    }

    func clearAll() {
        markers.values.forEach { $0.map = nil }
        markers.removeAll()
        circles.values.forEach { $0.map = nil }
        circles.removeAll()
        polylines.values.forEach { $0.map = nil }
        polylines.removeAll()
        polygons.values.forEach { $0.map = nil }
        polygons.removeAll()
    }

    func showMarkerInfoWindow(id: String) {
        markers[id]?.map = mapView
        mapView.selectedMarker = markers[id]
    }

    func hideMarkerInfoWindow(id: String) {
        if mapView.selectedMarker == markers[id] {
            mapView.selectedMarker = nil
        }
    }

    func isMarkerInfoWindowShown(id: String) -> Bool {
        return mapView.selectedMarker == markers[id]
    }
}
